File: Suyeon/ch23_DP/53_Maximum_Subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        for i in range(1,len(nums)):
            nums[i] += nums[i-1] if nums[i-1] > 0 else 0   # 새로운것에서 그전 까지의 합이 음수이면 첫지점 새로운것으로 다시 맞추기
        return max(nums)

File: Sujin/Ch23_Dynamic_Programming/53_Maximum_Subarray.cpp
class Solution {
public:
    int dp[100001]={0, };
    
    int maxSubArray(vector<int>& nums) {
        dp[0]=nums[0];
        int max_sum = dp[0];
        
        for(int i=1; i<nums.size(); i++){
            dp[i] = max(nums[i], dp[i-1]+nums[i]);
            max_sum = max(max_sum, dp[i]);
        }
        
        return max_sum;
        
    }
};​